lib/go-atscfg/parentdotconfig.go

```diff
@@ -405,17 +338,119 @@ func MakeParentDotConfig(
+
+	rankedParents := ParentInfoSortByRank(parentInfos)
+	sort.Sort(rankedParents)
+
+	parentInfo := []string{}
+	secondaryParentInfo := []string{}
+	nullParentInfo := []string{}
+	for _, parent := range ([]ParentInfo)(rankedParents) {
+		if !HasRequiredCapabilities(parent.Capabilities, ds.RequiredCapabilities) {
+			continue
+		}
+
+		if parent.PrimaryParent {
+			parentInfo = append(parentInfo, parent.Format())
+		} else if parent.SecondaryParent {
+			secondaryParentInfo = append(secondaryParentInfo, parent.Format())
+		} else {
+			nullParentInfo = append(nullParentInfo, parent.Format())
+		}
+	}
+
+	if len(parentInfo) == 0 {
+		// If no parents are found in the secondary parent either, then set the null parent list (parents in neither secondary or primary)
+		// as the secondary parent list and clear the null parent list.
+		if len(secondaryParentInfo) == 0 {
+			secondaryParentInfo = nullParentInfo
+			nullParentInfo = []string{}
+		}
+		parentInfo = secondaryParentInfo
+		secondaryParentInfo = []string{} // TODO should thi be '= secondary'? Currently emulates Perl
+	}
+
+	// TODO benchmark, verify this isn't slow. if it is, it could easily be made faster
+	seen := map[string]struct{}{} // TODO change to host+port? host isn't unique
+	parentInfo, seen = util.RemoveStrDuplicates(parentInfo, seen)
+	secondaryParentInfo, seen = util.RemoveStrDuplicates(secondaryParentInfo, seen)
+	nullParentInfo, seen = util.RemoveStrDuplicates(nullParentInfo, seen)
+
+	secondaryParentStr := strings.Join(secondaryParentInfo, "") + strings.Join(nullParentInfo, "")
+
+	// If the ats version supports it and the algorithm is consistent hash, put secondary and non-primary parents into secondary parent group.
+	// This will ensure that secondary and tertiary parents will be unused unless all hosts in the primary group are unavailable.
+
+	parents := ""
+	secondaryParents := ""
+
+	if atsMajorVer >= 6 && ds.MSOAlgorithm == "consistent_hash" && len(secondaryParents) > 0 {
+		parents = `parent="` + strings.Join(parentInfo, "") + `"`
+		secondaryParents = `" secondary_parent="` + secondaryParentStr + `"`
+	} else {
+		parents = `parent="` + strings.Join(parentInfo, "") + secondaryParentStr + `"`
+	}
+	return parents, secondaryParents
+}
+
 func MakeParentInfo(
 	server *ServerInfo,
 	serverDomain string, // getCDNDomainByProfileID(tx, server.ProfileID)
 	profileCaches map[ProfileID]ProfileCache, // getServerParentCacheGroupProfiles(tx, server)
-	originServers map[string][]CGServer, // getServerParentCacheGroupProfiles(tx, server)
-) map[string][]ParentInfo {
-	parentInfos := map[string][]ParentInfo{}
+	originServers map[OriginHost][]CGServer, // getServerParentCacheGroupProfiles(tx, server)
+) map[OriginHost][]ParentInfo {
+	parentInfos := map[OriginHost][]ParentInfo{}
 
 	// note servers also contains an "all" key
 	// originFQDN is "prefix" in Perl; ds is not really a "ds", that's what it's named in Perl
```

Comment:
Does `originFQDN` in the comment need updating?

Reply:
I removed it. I think it's diverged from the old Perl enough to no longer apply.